var/spack/repos/builtin/packages/dray/package.py

```diff
@@ -0,0 +1,332 @@
+    depends_on("apcomp+shared~openmp+mpi", when="+shared~openmp+mpi")
+    depends_on("apcomp~shared+openmp~mpi", when="~shared+openmp~mpi")
+    depends_on("apcomp+shared+openmp~mpi", when="+shared+openmp~mpi")
+    depends_on("apcomp~shared~openmp~mpi", when="~shared~openmp~mpi")
+    depends_on("apcomp+shared~openmp~mpi", when="+shared~openmp~mpi")
+
+    depends_on("raja@0.9.0+cuda~openmp+shared", when="+cuda~openmp+shared")
+    depends_on("raja@0.9.0+cuda+openmp+shared", when="+cuda+openmp+shared")
+    depends_on("raja@0.9.0+cuda~openmp~shared", when="+cuda~openmp~shared")
+    depends_on("raja@0.9.0+cuda+openmp~shared", when="+cuda+openmp~shared")
+
+    depends_on("raja@0.9.0~cuda~openmp+shared", when="~cuda~openmp+shared")
+    depends_on("raja@0.9.0~cuda+openmp+shared", when="~cuda+openmp+shared")
+    depends_on("raja@0.9.0~cuda~openmp~shared", when="~cuda~openmp~shared")
+    depends_on("raja@0.9.0~cuda+openmp~shared", when="~cuda+openmp~shared")
+
+    depends_on("umpire@1.0.0+cuda+shared", when="+cuda+shared")
+    depends_on("umpire@1.0.0+cuda~shared", when="+cuda~shared")
+    depends_on("umpire@1.0.0~cuda+shared", when="~cuda+shared")
+    depends_on("umpire@1.0.0~cuda~shared", when="~cuda~shared")
+
+    depends_on("mfem+shared+conduit~threadsafe", when="+shared")
+    depends_on("mfem~shared+conduit~threadsafe", when="~shared")
+
+    def setup_build_environment(self, env):
+        env.set('CTEST_OUTPUT_ON_FAILURE', '1')
+
+    def install(self, spec, prefix):
+        """
+        Build and install Devil Ray.
+        """
+        with working_dir('spack-build', create=True):
+            py_site_pkgs_dir = None
+
+            host_cfg_fname = self.create_host_config(spec,
+                                                     prefix,
+                                                     py_site_pkgs_dir)
+            cmake_args = []
+            # if we have a static build, we need to avoid any of
+            # spack's default cmake settings related to rpaths
+            # (see: https://github.com/LLNL/spack/issues/2658)
+            if "+shared" in spec:
+                cmake_args.extend(std_cmake_args)
+            else:
+                for arg in std_cmake_args:
+                    if arg.count("RPATH") == 0:
+                        cmake_args.append(arg)
+            cmake_args.extend(["-C", host_cfg_fname, "../src"])
+            print("Configuring Devil Ray...")
+            cmake(*cmake_args)
+            print("Building Devil Ray...")
+            make()
+            # run unit tests if requested
+            if "+test" in spec and self.run_tests:
+                print("Running Devil Ray Unit Tests...")
+                make("test")
+            print("Installing Devil Ray...")
+            make("install")
+            # install copy of host config for provenance
+            install(host_cfg_fname, prefix)
+
+    def create_host_config(self, spec, prefix, py_site_pkgs_dir=None):
```

Comment:
Is there any reason we can't extend `CMakePackage` like every other Spack package so we don't need all of this custom logic?

Reply:
Yes. One reason, of several, is that we often need to create a build of our packages outside of spack, and re-using the deps that spack build. This is a way of outputting those deps. It also adds some transparency into exactly how the package was built so that we can debug issues.

Reply:
Another reason is that its an easy way to give someone, who doesn't know spack, a way to build the library on systems where we have already have a spack install.

Reply:
Perhaps we should be doing this in the CMakePackage base class then? Then each package only needs to provide `cmake_args` and the base class will take care of the rest. I like things to be uniform.

Reply:
While I do advocate for this approach, I am not sure it makes sense to impose this style on other projects. 
It's a process we follow for several projects and it is important to our development workflow. (We have used it for ~5 years) 

How to capture seems like a something for a broader spack discussion to address.

Reply:
@cyrush can you (or someone else) get that discussion started? Just open a new issue so we don't forget about this. I'm fine with different strategies but I would like all Spack packages to be handled uniformly.